provider/azurerm: add virtual_machine_extension resource

pmcatominey opened the pull request:
Picked up from where #6548 left off

Fixes #5820 

settings and protected_settings take JSON objects as strings to make extension
generic

```
TF_ACC=1 go test ./builtin/providers/azurerm -v -run "TestAccAzureRMVirtualMachineExtension" -timeout 120m
=== RUN   TestAccAzureRMVirtualMachineExtension_importBasic
--- PASS: TestAccAzureRMVirtualMachineExtension_importBasic (810.44s)
=== RUN   TestAccAzureRMVirtualMachineExtension_basic
--- PASS: TestAccAzureRMVirtualMachineExtension_basic (839.55s)
=== RUN   TestAccAzureRMVirtualMachineExtension_concurrent
--- PASS: TestAccAzureRMVirtualMachineExtension_concurrent (689.69s)
PASS
ok          github.com/hashicorp/terraform/builtin/providers/azurerm        2339.695s
```

stack72 commented:
Hi @pmcatominey 

Looking at this right now - what are your thoughts on adding a test for each of the Extension types listed in this [document](https://azure.microsoft.com/en-gb/documentation/articles/virtual-machines-windows-extensions-configuration-samples/)? This would guarantee we can support the main types

P.

stack72 commented on builtin/providers/azurerm/resource_arm_virtual_machine_extension.go:
```diff
@@ -0,0 +1,232 @@
+	d.Set("type_handler_version", resp.Properties.TypeHandlerVersion)
+	d.Set("auto_upgrade_minor_version", resp.Properties.AutoUpgradeMinorVersion)
+
+	settings, err := flattenArmVirtualMachineExtensionSettings(*resp.Properties.Settings)
```
Is there *any* change that Settings could be nil? If so, then we are going to cause a panic. Worth passing the Settings struct as a pointer to the flatten func and carrying out the necessary checks in there?

pmcatominey replied:
Seems that `settings` is optional as far as the API is concerned, I thought it was required, I'll guard against a panic in the read since it's the only settings map read back.

stack72 commented on builtin/providers/azurerm/resource_arm_virtual_machine_extension_test.go:
I'd love to see a more complex example and perhaps something with ProtectedSettings as well to make sure that it works

pmcatominey commented:
Updated the handling of settings as it is an optional field, added a test using the Linux Diagnostics extension to cover usage of protected_settings:

```
TF_ACC=1 go test ./builtin/providers/azurerm -v -run "TestAccAzureRMVirtualMachineExtension" -timeout 120m
=== RUN   TestAccAzureRMVirtualMachineExtension_importBasic
--- PASS: TestAccAzureRMVirtualMachineExtension_importBasic (697.55s)
=== RUN   TestAccAzureRMVirtualMachineExtension_basic
--- PASS: TestAccAzureRMVirtualMachineExtension_basic (824.17s)
=== RUN   TestAccAzureRMVirtualMachineExtension_concurrent
--- PASS: TestAccAzureRMVirtualMachineExtension_concurrent (929.74s)
=== RUN   TestAccAzureRMVirtualMachineExtension_linuxDiagnostics
--- PASS: TestAccAzureRMVirtualMachineExtension_linuxDiagnostics (803.19s)
PASS
ok 	github.com/hashicorp/terraform/builtin/providers/azurerm	3254.663s
```

stack72 commented:
Thanks for the additions! This LGTM :)
